From 1ce644c457b692d69384d8a3d4a64cb34ffb3d9d Mon Sep 17 00:00:00 2001
From: NoisyCoil <noisycoil@tutanota.com>
Date: Tue, 11 Feb 2025 13:24:15 +0100
Subject: [PATCH] Backport unsafe_op_in_unsafe_fn fixes

Adapted from:
- 068d4ce0da69 Fixes #3123
- ee2a2acc365e Add missed unsafe in the raw_set_bit function
---
 codegen/bitfield_unit_raw_ref_macros.rs | 16 +++++++++-------
 1 file changed, 9 insertions(+), 7 deletions(-)

diff --git a/codegen/bitfield_unit_raw_ref_macros.rs b/codegen/bitfield_unit_raw_ref_macros.rs
index 3411c22e..0c864c73 100644
--- a/codegen/bitfield_unit_raw_ref_macros.rs
+++ b/codegen/bitfield_unit_raw_ref_macros.rs
@@ -43,8 +43,8 @@ where
         debug_assert!(index / 8 < core::mem::size_of::<Storage>());
 
         let byte_index = index / 8;
-        let byte = *(core::ptr::addr_of!((*this).storage) as *const u8)
-            .offset(byte_index as isize);
+        let byte = unsafe { *(core::ptr::addr_of!((*this).storage) as *const u8)
+            .offset(byte_index as isize) };
 
         Self::extract_bit(byte, index)
     }
@@ -80,10 +80,12 @@ where
         debug_assert!(index / 8 < core::mem::size_of::<Storage>());
 
         let byte_index = index / 8;
-        let byte = (core::ptr::addr_of_mut!((*this).storage) as *mut u8)
-            .offset(byte_index as isize);
+        let byte = unsafe {
+            (core::ptr::addr_of_mut!((*this).storage) as *mut u8)
+                .offset(byte_index as isize)
+        };
 
-        *byte = Self::change_bit(*byte, index, val);
+        unsafe { *byte = Self::change_bit(*byte, index, val) };
     }
 
     #[inline]
@@ -127,7 +129,7 @@ where
         let mut val = 0;
 
         for i in 0..(bit_width as usize) {
-            if Self::raw_get_bit(this, i + bit_offset) {
+            if unsafe { Self::raw_get_bit(this, i + bit_offset) } {
                 let index = if cfg!(target_endian = "big") {
                     bit_width as usize - 1 - i
                 } else {
@@ -183,7 +185,7 @@ where
             } else {
                 i
             };
-            Self::raw_set_bit(this, index + bit_offset, val_bit_is_set);
+            unsafe { Self::raw_set_bit(this, index + bit_offset, val_bit_is_set) };
         }
     }
 }
-- 
2.47.2

